### PR TITLE
Fix the disconnect between namespaces on install

### PR DESCRIPTION
**Pull request by Itxaka**
This re-Inits the config so we can set the Storage Driver
namespace to what is obtained from the chart.

This is because on helm you can only set the namespace via
the --namespace flag which gets into the envsettings.namespace
and the helm.go loads that var into the Configuration object
during cobra.OnInitialize.

To maintain compatibility we do the same but then during
the command execution we read the chart annotations and
change the release namespace on the fly.
This is ok for the namespace install but the Storage Driver,
which is already loaded, has a different namespace
(usually default) and the Release gets stored in the wrong
namespace (stored *for* helm, for k8s is installed properly)
This results into a disconnection between where the release
is installed and where hypper/helm looks in the Storage Driver
to search for it.

With this below we re-Init the configuration and hance the
Storage Driver to have the namespace correctly set in sync
with the Release

Maybe fixes: #75 ???

Signed-off-by: Itxaka <igarcia@suse.com>

### CODE REVIEW

**Comment by Itxaka** [comment]
test are failing, I know, first I want to see if this is a viable solution. I do not understand the full implications of re-Init the config. Locally it seems ok, I could not find anything that was missed by re-initializing but I rather have @mattfarina have a good look and maybe offer an alternative if its simpler.


BTW some tests seems to be fail now like ....properly. Not able to install to X namespace becuase the namespace doesnt exists seems like the proper outcome, so this may have fixed the issue actually and the tests were just installing into the default namespace!

also some upgrade test which I probably broke or was broken since day one :clown_face:

File: cmd/hypper/hypper.go [hunk 1]
```diff
@@ -114,3 +114,18 @@ func loadReleasesInMemory(actionConfig *action.Configuration) {
...
+// the envsettings.namespace and the helm.go loads that var into the Configuration object during
+// cobra.OnInitialize. To maintain compativility we do the same but then during the command execution
+// we read the chart annotations and change the release namespace on the fly. This is ok for the
+// namespace install but the Storage Driver, which is already loaded, has a different namespace (usually default)
+// and the Release gets stored in the wrong namespace (stored *for* helm, for k8s is installed properly)
+// This results into a disconnect between where the release is installed and where hypper/helm looks in the
+// Storage Driver to search for it. With this below we re-Init the configuration and hance the Storage Driver to
+// have the namespace correctly set in sync with the Release
+func ReinitConfigForNamespace(cfg *action.Configuration, namespace string, logger log.Logger) {
+	_ = cfg.Init(settings.RESTClientGetter(), namespace, os.Getenv("HELM_DRIVER"), logger.Debugf)
```
Comment (dragonchaser) [inline cmd/hypper/hypper.go:130]: Reading env variables outside environment.go... I do not really like that, can we move that to a settings variable?

Just saw this is also down when assigning the `helmDriver` variable in line 65 and in `cmd/hypper/list.go`.....

**Review by dragonchaser — CHANGES_REQUESTED** [review]

File: cmd/hypper/hypper.go (hunk 1 above)
Comment (Itxaka) [reply cmd/hypper/hypper.go:130]: well, I was trying to mimick the first `Init` which is done just above in line 65. This is how helm does it, so Im guessing we just lift it from there?

Comment (Itxaka) [reply cmd/hypper/hypper.go:130]: I wonder why they dont directly use `settings` here to set both of those options...maybe an oversight??

Comment (Itxaka) [reply cmd/hypper/hypper.go]: updated to use settings as it turns out that we can add it there with no issues that I can see. I wonder if we should upstream this and the other env var used...

File: cmd/hypper/install.go [hunk 2]
```diff
@@ -120,6 +119,9 @@ func runInstall(args []string, client *action.Install, valueOpts *values.Options
 		client.SetNamespace(chartRequested, settings.Namespace())
 	}
 
+	// Reinit the config so we set the proper namespace of the Storage Driver
+	ReinitConfigForNamespace(client.Config, client.Namespace, logger)
```
Comment (viccuad) [inline cmd/hypper/install.go:123]: Why not directly do `client.Config.Init(settings.RESTClientGetter(), client.Namespace, settings.HelmDriver, logger.Debugf)`?
I find it more readable that a static function that contains no logic, and no changes in the scope of vars/functions it access.

File: cmd/hypper/hypper.go [hunk 3]
```diff
@@ -114,3 +114,18 @@ func loadReleasesInMemory(actionConfig *action.Configuration) {
 	// Must reset namespace to the proper one
 	mem.SetNamespace(settings.Namespace())
 }
+
+// This re-Inits the config so we can set the Storage Driver namespace to what is obtained
```
Comment (viccuad) [inline cmd/hypper/hypper.go:118]: This comment is super useful, thanks! :D

nit: But I think it's maybe to big. I would prefer if it were on the commit message of the commit that adds the fix, accessible from git blame.

File: pkg/cli/environment.go [hunk 4]
```diff
@@ -95,6 +96,7 @@ func New() *EnvSettings {
 		RegistryConfig:   envOr("HYPPER_REGISTRY_CONFIG", hypperpath.ConfigPath("registry.json")),
 		RepositoryConfig: envOr("HYPPER_REPOSITORY_CONFIG", hypperpath.ConfigPath("repositories.yaml")),
 		RepositoryCache:  envOr("HYPPER_REPOSITORY_CACHE", hypperpath.CachePath("repository")),
+		HelmDriver:       os.Getenv("HELM_DRIVER"),
```
Comment (viccuad) [inline pkg/cli/environment.go:99]: :100: !

**Review by viccuad — CHANGES_REQUESTED** [review]
Wooho! good find! :D
And the helmDriver change is also a win!

File: cmd/hypper/install.go (hunk 2 above)
Comment (viccuad) [reply cmd/hypper/install.go:123]: Shouldn't we have access to client.Config.Init there, too?

File: cmd/hypper/hypper.go (hunk 3 above)
Comment (viccuad) [reply cmd/hypper/hypper.go:118]: I like the one-line comment `// Reinit the config so we set the proper namespace of the Storage Driver` when it's called, with that is obvious that the call has a purpose. But I'm ok either way, it's a nit :).

File: cmd/hypper/install.go (hunk 2 above)
Comment (Itxaka) [reply cmd/hypper/install.go:123]: we should but if we are gonna reuse this in teh upgrade for example I rather have a common function that sets this.

In any case this does not work as expected as it clears the full KubeClient ¬_¬

**Comment by Itxaka** [comment]
Unfortunately, not a good solution as we wipe out a lot of stuff :(

File: pkg/cli/environment.go (hunk 4 above)
Comment (viccuad) [reply pkg/cli/environment.go:99]: ok, I don't like changing `HELM_DRIVER` to `HYPPER_DRIVER` anymore. If we change it, we need to also start overloading `ActionConfig.Init()` and wherever it gets used as `HELM_DRIVER`, to `HYPPER_DRIVER`. Example:

https://github.com/helm/helm/blob/6e4a61f21279560b4a4cf7b84088c9e7e75757a9/pkg/action/action.go#L411

Comment (dragonchaser) [reply pkg/cli/environment.go:99]: Yes but this calls of implementing #65 urgently so we do not miss any changes there....

**Pull request closed without merging**